File: src/GenAIRR/_native/csrc/benchmarks/bench_e2e.c
/**
 * bench_e2e.c — End-to-end benchmark: rearrange + S5F + serialize.
 *
 * Usage: ./bench_e2e [N]
 *   N = number of sequences to generate (default 10000)
 *
 * Measures wall-clock time for:
 *   Phase 1: Config + correction context setup
 *   Phase 2: N × (pipeline_execute + s5f_mutate + airr_serialize)
 *   Phase 3: Write all records to TSV file
 *
 * Output: timing breakdown + summary statistics.
 */

#include "human_igh_imgt.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static double now_sec(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}

int main(int argc, char **argv) {
    int N = 10000;
    if (argc > 1) N = atoi(argv[1]);
    if (N <= 0) N = 10000;

    printf("GenAIRR-C End-to-End Benchmark\n");
    printf("==============================\n");
    printf("Generating %d sequences (HUMAN_IGH_IMGT + S5F)\n\n", N);

    /* ── Phase 1: Setup ──────────────────────────────────────── */
    double t0 = now_sec();

    SimConfig cfg;
    human_igh_imgt_load_config(&cfg);

    S5FModel s5f;
    s5f_model_init(&s5f, 0.05, 0.15, false);
    human_igh_imgt_load_s5f(&s5f);

    Pipeline pl = pipeline_build(&cfg);

    AlleleCorrectionSet ctx = allele_correction_set_build(&cfg);

    double t_setup = now_sec() - t0;
    printf("Phase 1 — Setup:            %8.3f ms\n", t_setup * 1000.0);
    printf("  V alleles: %d, D alleles: %d, J alleles: %d\n",
           cfg.v_alleles.count, cfg.d_alleles.count, cfg.j_alleles.count);
    printf("  Short-D threshold: %d\n\n", ctx.d_bitmap.short_d_threshold);

    /* ── Phase 2: Generate sequences ─────────────────────────── */
    AirrRecord *records = calloc(N, sizeof(AirrRecord));
    if (!records) {
        fprintf(stderr, "Failed to allocate %d records\n", N);
        return 1;
    }

    RngState bench_rng;
    rng_seed(&bench_rng, 42, 0);
    cfg.rng = &bench_rng;

    double t1 = now_sec();

    int success = 0;
    int total_len = 0;
    int total_mutations = 0;
    int total_productive = 0;
    double total_mutation_rate = 0.0;

    for (int i = 0; i < N; i++) {
        ASeq seq;
        aseq_init(&seq);
        SimRecord rec;
        sim_record_init(&rec);

        /* Rearrange */
        pipeline_execute(&pl, &cfg, &seq, &rec, 0, NULL, NULL);
        if (seq.length == 0) {
            aseq_reset(&seq);
            continue;
        }

        /* Mutate */
        S5FResult mres;
        s5f_mutate(&s5f, &seq, &rec, &bench_rng, &mres);

        /* Serialize */
        airr_serialize(&seq, &rec, &cfg, &ctx, &records[success]);

        total_len += records[success].sequence_length;
        total_mutations += records[success].n_mutations;
        total_mutation_rate += records[success].mutation_rate;
        if (records[success].productive) total_productive++;
        success++;

        aseq_reset(&seq);
    }

    double t_generate = now_sec() - t1;

    printf("Phase 2 — Generate:         %8.3f ms  (%d sequences)\n",
           t_generate * 1000.0, success);
    printf("  Per sequence:             %8.3f µs\n",
           t_generate / success * 1e6);
    printf("  Throughput:               %8.0f seq/sec\n",
           success / t_generate);
    printf("  Avg length:               %8.1f bp\n",
           (double)total_len / success);
    printf("  Avg mutations:            %8.1f per seq\n",
           (double)total_mutations / success);
    printf("  Avg mutation rate:        %8.4f\n",
           total_mutation_rate / success);
    printf("  Productive:               %8d / %d (%.1f%%)\n\n",
           total_productive, success,
           100.0 * total_productive / success);

    /* ── Phase 3: Write TSV ──────────────────────────────────── */
    double t2 = now_sec();

    FILE *fp = fopen("/tmp/genairr_bench.tsv", "w");
    if (fp) {
        airr_write_tsv_header(fp);
        for (int i = 0; i < success; i++) {
            airr_write_tsv_row(fp, &records[i]);
        }
        fclose(fp);
    }

    double t_write = now_sec() - t2;
    printf("Phase 3 — Write TSV:        %8.3f ms\n", t_write * 1000.0);

    /* ── Summary ─────────────────────────────────────────────── */
    double t_total = t_setup + t_generate + t_write;
    printf("\n");
    printf("Total wall time:            %8.3f ms\n", t_total * 1000.0);
    printf("  Setup:                    %8.1f%%\n", 100.0 * t_setup / t_total);
    printf("  Generate:                 %8.1f%%\n", 100.0 * t_generate / t_total);
    printf("  Write:                    %8.1f%%\n", 100.0 * t_write / t_total);

    if (fp) {
        /* Get file size */
        fp = fopen("/tmp/genairr_bench.tsv", "r");
        if (fp) {
            fseek(fp, 0, SEEK_END);
            long size = ftell(fp);
            fclose(fp);
            printf("\nOutput: /tmp/genairr_bench.tsv (%.1f MB, %d rows)\n",
                   size / 1e6, success);
        }
    }

    free(records);
    allele_correction_set_destroy(&ctx);
    sim_config_destroy(&cfg);
    return 0;
}

File: src/GenAIRR/_native/csrc/benchmarks/bench_e2e.py
#!/usr/bin/env python3
"""
bench_e2e.py — Python benchmark equivalent for comparison with C.

Usage: python3 benchmarks/bench_e2e.py [N]
  N = number of sequences to generate (default 10000)

Measures wall-clock time for the same pipeline as bench_e2e.c:
  Phase 1: Load DataConfig + compile
  Phase 2: N × simulate (rearrange + S5F mutate)
  Phase 3: Convert to AIRR dicts
"""

import sys
import time
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from GenAIRR import Experiment


def main():
    N = int(sys.argv[1]) if len(sys.argv) > 1 else 10000

    print("GenAIRR Python End-to-End Benchmark")
    print("====================================")
    print(f"Generating {N} sequences (HUMAN_IGH + S5F)\n")

    # Phase 1: Setup
    t0 = time.perf_counter()
    from GenAIRR.ops import rate, model
    sim = (Experiment.on("human_igh")
           .mutate(rate(0.05, 0.15), model("s5f"))
           .compile(seed=42))
    t_setup = time.perf_counter() - t0
    print(f"Phase 1 — Compile:          {t_setup*1000:8.3f} ms")

    # Phase 2: Generate N sequences
    t1 = time.perf_counter()
    result = sim.simulate(n=N)
    records = list(result)
    t_generate = time.perf_counter() - t1

    print(f"Phase 2 — Generate:         {t_generate*1000:8.3f} ms  ({len(records)} sequences)")
    print(f"  Per sequence:             {t_generate/len(records)*1e6:8.3f} µs")
    print(f"  Throughput:               {len(records)/t_generate:8.0f} seq/sec")

    # Compute stats
    lengths = [len(r.get('sequence', '')) for r in records]
    mut_rates = [r.get('mutation_rate', 0) for r in records]
    productive = sum(1 for r in records if r.get('productive', False))

    avg_len = sum(lengths) / len(lengths)
    avg_rate = sum(mut_rates) / len(mut_rates)

    print(f"  Avg length:               {avg_len:8.1f} bp")
    print(f"  Avg mutation rate:        {avg_rate:8.4f}")
    print(f"  Productive:               {productive:8d} / {len(records)} ({100*productive/len(records):.1f}%)")

    # Phase 3: Write TSV
    t2 = time.perf_counter()
    outpath = "/tmp/genairr_bench_py.tsv"
    import csv
    if records:
        keys = records[0].keys()
        with open(outpath, 'w', newline='') as f:
            w = csv.DictWriter(f, fieldnames=keys, delimiter='\t')
            w.writeheader()
            w.writerows(records)
    t_write = time.perf_counter() - t2
    print(f"\nPhase 3 — Write TSV:        {t_write*1000:8.3f} ms")

    t_total = t_setup + t_generate + t_write
    print(f"\nTotal wall time:            {t_total*1000:8.3f} ms")
    print(f"  Compile:                  {100*t_setup/t_total:8.1f}%")
    print(f"  Generate:                 {100*t_generate/t_total:8.1f}%")
    print(f"  Write:                    {100*t_write/t_total:8.1f}%")

    fsize = os.path.getsize(outpath) if os.path.exists(outpath) else 0
    print(f"\nOutput: {outpath} ({fsize/1e6:.1f} MB, {len(records)} rows)")


if __name__ == "__main__":
    main()
